### DISCUSSION TITLE
Manual/ExecutionOrder

### DISCUSSION THREAD

**Discussion by docscomments** [Developer-Notes-Extension/Unity-Notes#288] [Notes]
# Manual/ExecutionOrder

https://docs.unity3d.com/Manual/ExecutionOrder.html

**Comment by docscomments** [comment]
**Comment by `@HilariousCow`:**  

Constraints systems fire last thing before LateUpdate.  
Also missing: ProcessRootMotion (part of IAnimationJob) is executed just before OnAnimatorMove, but after ProcessGraph (e.g. "PrepareFrame" calls to your ScriptablePlayables).

**Comment by PeturDarri** [comment]
These are some undocumented event functions that I found used in various editor classes and in the Timeline package:
```cs
void __internalAwake();
void OnEnableINTERNAL();
void OnDisableINTERNAL();
```
Here's how their order compares to their public counterparts (and `OnDestroy`):
```
__internalAwake
Awake
OnEnableINTERNAL
OnEnable
OnDisable
OnDisableINTERNAL
OnDestroy
```
These appear to be used by Unity to ensure that initialization can occur before the user's code and finalization occurs after.
